### PR TITLE
Remove dependence of aries module on VcxStateType

### PR DESCRIPTION
**Pull request by mirgee**
This PR adds new enums representing current state of `IssuerSM`, `HolderSM`, `ProverSM` and `VerifierSM`: `IssuerState`, `HolderState`, `ProverState` and `VerifierState` respectively. Function `get_state()` returns variants of said enums on the internal `aries` level, and these are converted to numeric representations on the C-callable level.

Wrapper now exports enums representing these states in place of `StateType`: `IssuerStateType`, `HolderStateType`, `VerifierStateType`, `IssuerStateType`, `ConnectionStateType`.

This is a breaking change, since the variant representations corresponding to the states returned by `get_state()` function has changed, as well as the numeric representations. For the numeric values, see [here](https://github.com/hyperledger/aries-vcx/pull/314/files#diff-dd5350e4ee5378b3a042f96e8a6fe1cfe838a8a937b55104953acd10d83d8040R34-R80), and for the changes to `get_state()` return values, see corresponding implementations for each of the concerned SMs.

Signed-off-by: Miroslav Kovar <miroslavkovar@protonmail.com>

### CODE REVIEW

**Comment by Patrik-Stas** [comment]
@mirgee the failing tests seems to indicate that the state mapping is a bit off somewhere

File: agents/node/vcxagent-core/src/services/service-connections.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ module.exports.createServiceConnections = function createServiceConnections ({ l
 
   async function _progressConnectionToAcceptedState (connection, attemptsThreshold, timeoutMs) {
     async function progressToAcceptedState () {
-      if (await connection.updateState() !== StateType.Accepted) {
+      if (await connection.updateState() !== InviterStateType.Completed) {
```
Comment (Patrik-Stas) [inline agents/node/vcxagent-core/src/services/service-connections.js:53]: great to have the state constants separated betweem different SMs 👍

File: agents/node/vcxagent-core/src/services/service-cred-holder.js [hunk 2]
```diff
@@ -49,7 +50,7 @@ module.exports.createServiceCredHolder = function createServiceCredHolder ({ log
   async function waitForCredential (connectionId, holderCredentialId, attemptsThreshold = 10, timeoutMs = 2000) {
     const connection = await loadConnection(connectionId)
     const credential = await loadHolderCredential(holderCredentialId)
-    await _progressCredentialToState(credential, connection, StateType.Accepted, attemptsThreshold, timeoutMs)
+    await _progressCredentialToState(credential, connection, 2, attemptsThreshold, timeoutMs)
```
Comment (Patrik-Stas) [inline agents/node/vcxagent-core/src/services/service-cred-holder.js:53]: forgotten numeric constant

File: ci/indy-pool.dockerfile [hunk 3]
```diff
@@ -33,7 +33,7 @@ ARG python3_orderedset_ver=2.0
 ARG python3_psutil_ver=5.4.3
 ARG python3_pympler_ver=0.5
 
-RUN apt-get update -y && apt-get install -y \
+RUN apt-get update -y && apt-get install -y --allow-unauthenticated \
```
Comment (Patrik-Stas) [inline ci/indy-pool.dockerfile:36]: some issues with the docker build?

File: libvcx/src/api_lib/api_c/credential.rs [hunk 4]
```diff
@@ -1070,7 +1069,7 @@ mod tests {
         let _setup = SetupMocks::init();
 
         let handle = credential::credential_create_with_offer("test_send_request", ARIES_CREDENTIAL_OFFER).unwrap();
-        assert_eq!(credential::get_state(handle).unwrap(), VcxStateType::VcxStateRequestReceived as u32);
+        assert_eq!(credential::get_state(handle).unwrap(), HolderState::OfferReceived as u32);
```
Comment (Patrik-Stas) [inline libvcx/src/api_lib/api_c/credential.rs:1072]: Let's improve this test and add state assert after wee call `vcx_credential_send_request`

File: ci/indy-pool.dockerfile (hunk 3 above)
Comment (mirgee) [reply ci/indy-pool.dockerfile:36]: https://github.com/hyperledger/indy-sdk/issues/2407

File: libvcx/src/api_lib/api_c/credential.rs (hunk 4 above)
Comment (Patrik-Stas) [reply libvcx/src/api_lib/api_c/credential.rs:1072]: Actually maybe we can just delete this test. This test is subset of of  `test_vcx_credential_update_state`

File: libvcx/src/api_lib/api_c/issuer_credential.rs [hunk 5]
```diff
@@ -94,16 +93,7 @@ pub extern fn vcx_issuer_create_credential(command_handle: CommandHandle,
...
-    let issuer_did: String = if !issuer_did.is_null() {
-        check_useful_c_str!(issuer_did, VcxErrorKind::InvalidOption);
-        issuer_did.to_owned()
-    } else {
-        match settings::get_config_value(settings::CONFIG_INSTITUTION_DID) {
-            Ok(x) => x,
-            Err(x) => return x.into()
-        }
-    };
+    check_useful_c_str!(issuer_did, VcxErrorKind::InvalidOption);
```
Comment (Patrik-Stas) [inline libvcx/src/api_lib/api_c/issuer_credential.rs:96]: Don't forget to also include this breaking change in PR description

File: libvcx/src/aries/handlers/issuance/holder/holder.rs [hunk 6]
```diff
@@ -84,14 +92,14 @@ impl Holder {
 
     pub fn update_state(&mut self, connection: &Connection) -> VcxResult<u32> {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/issuance/holder/holder.rs:93]: should return `VcxResult<HolderState>`

File: libvcx/src/aries/handlers/issuance/holder/state_machine.rs [hunk 7]
```diff
@@ -42,14 +42,14 @@ impl HolderSM {
...
-            HolderState::RequestSent(_) => VcxStateType::VcxStateOfferSent as u32,
-            HolderState::Finished(ref status) => {
+            HolderFullState::OfferReceived(_) => HolderState::OfferReceived,
+            HolderFullState::RequestSent(_) => HolderState::RequestSent,
+            HolderFullState::Finished(ref status) => {
                 match status.status {
-                    Status::Success => VcxStateType::VcxStateAccepted as u32,
-                    _ => VcxStateType::VcxStateNone as u32,
+                    Status::Success => HolderState::Finished,
+                    _ => HolderState::Failed
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/issuance/holder/state_machine.rs:52]: Please include numeric state changes in PR description (like here, `VcxStateNone` was none, now we have `HolderState::Failed` which is `3`

File: libvcx/src/aries/handlers/issuance/holder/state_machine.rs [hunk 8]
```diff
@@ -60,10 +60,10 @@ impl HolderSM {
 
         for (uid, message) in messages {
             match self.state {
-                HolderState::OfferReceived(_) => {
+                HolderFullState::OfferReceived(_) => {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/issuance/holder/state_machine.rs:63]: This code can look slightly cleaner if we match against `self.get_state()` instead

File: libvcx/src/aries/handlers/issuance/issuer/issuer.rs [hunk 9]
```diff
@@ -18,6 +18,16 @@ pub struct IssuerConfig {
     pub tails_file: Option<String>
 }
 
+#[derive(Debug, PartialEq)]
+pub enum IssuerState {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/issuance/issuer/issuer.rs:22]: I think this should be defined in `state_machine.rs`

File: libvcx/src/aries/handlers/issuance/issuer/state_machine.rs [hunk 10]
```diff
@@ -179,16 +179,16 @@ impl IssuerSM {
         None
     }
 
-    pub fn state(&self) -> u32 {
+    pub fn get_state(&self) -> IssuerState {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/issuance/issuer/state_machine.rs:182]: I see also couple changes here which projects into breaking changes on numeric level - please include in PR description. Not sure if it's worthy to go into details if which state number changed exactly how, but at least note that state mapping has changed for this FSM.

File: libvcx/src/aries/handlers/proof_presentation/prover/prover.rs [hunk 11]
```diff
@@ -15,6 +15,16 @@ pub struct Prover {
     prover_sm: ProverSM,
 }
 
+#[derive(Debug, PartialEq)]
+pub enum ProverState {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/proof_presentation/prover/prover.rs:19]: This should also be moved to on `state_machine.rs` level

File: libvcx/src/aries/handlers/proof_presentation/verifier/state_machine.rs [hunk 12]
```diff
@@ -44,10 +43,10 @@ impl VerifierSM {
 
         for (uid, message) in messages {
             match self.state {
-                VerifierState::Initiated(_) => {
+                VerifierFullState::Initiated(_) => {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/proof_presentation/verifier/state_machine.rs:46]: We could use `VerifierState` to make this code cleaner

File: libvcx/src/aries/handlers/proof_presentation/verifier/verifier.rs [hunk 13]
```diff
@@ -12,6 +11,14 @@ pub struct Verifier {
     verifier_sm: VerifierSM
 }
 
+#[derive(Debug, PartialEq)]
+pub enum VerifierState {
```
Comment (Patrik-Stas) [inline libvcx/src/aries/handlers/proof_presentation/verifier/verifier.rs:15]: analogous, move to `state_machine.rs`

File: libvcx/src/libindy/utils/wallet.rs [hunk 14]
```diff
@@ -544,7 +542,7 @@ pub mod tests {
         let id = "id1";
         let value = "value1";
 
-        api_c::vcx::vcx_shutdown(true);
+        // api_c::vcx::vcx_shutdown(true);
```
Comment (Patrik-Stas) [inline libvcx/src/libindy/utils/wallet.rs:545]: This test is currently in `to_restore` mode, is this change needed?

File: wrappers/node/test/suite1/ariesvcx-credential.test.ts [hunk 15]
```diff
@@ -94,9 +94,9 @@ describe('Credential:', () => {
       const connection = await createConnectionInviterRequested();
       const data = await dataCredentialCreateWithOffer();
       const credential = await Credential.create(data);
-      assert.equal(await credential.getState(), StateType.RequestReceived);
+      assert.equal(await credential.getState(), 0);
```
Comment (Patrik-Stas) [inline wrappers/node/test/suite1/ariesvcx-credential.test.ts:97]: Should use the TS enums rather than numeric values

File: Cargo.toml [hunk 16]
```diff
@@ -9,3 +9,6 @@ members = [
 debug = true
 panic = 'unwind'
 incremental = false
+ 
+[profile.dev]
+split-debuginfo = "unpacked"
```
Comment (Patrik-Stas) [inline Cargo.toml:14]: Does this comes with some performance benefits for debug compilation? Is there a practical advantage of `unpacked` VS other options?

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
Minor changes requested, overally love it! 👍

File: libvcx/src/aries/handlers/issuance/holder/state_machine.rs (hunk 8 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/issuance/holder/state_machine.rs:63]: That's true, I'll do that everywhere

File: libvcx/src/aries/handlers/issuance/holder/state_machine.rs (hunk 7 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/issuance/holder/state_machine.rs:52]: I am not sure, are these important? Any user should be using the constants / enums. The numeric values of these are subject to change anyway.

File: libvcx/src/aries/handlers/issuance/holder/holder.rs (hunk 6 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/issuance/holder/holder.rs:93]: That's true, good catch, don't know how I missed that

File: libvcx/src/aries/handlers/issuance/holder/state_machine.rs (hunk 8 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/issuance/holder/state_machine.rs:63]: Yes, but then it would be inconsistent with other SMs where the enums holds useful data.

File: Cargo.toml (hunk 16 above)
Comment (mirgee) [reply Cargo.toml:14]: Yes, it it said to enable faster debug builds on MacOS (it used to use a slow tool for incremental builds by default), but by [digging deeper](https://github.com/rust-lang/cargo/pull/9112) I found out that it should not be necessary with newer rustc versions anymore. 👍

File: libvcx/src/aries/handlers/proof_presentation/verifier/state_machine.rs (hunk 12 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/proof_presentation/verifier/state_machine.rs:46]: As I said above, I would prefer consistency / symmetry over such a minor improvement in readability.

File: libvcx/src/aries/handlers/issuance/issuer/issuer.rs (hunk 9 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/issuance/issuer/issuer.rs:22]: `state_machine` is a private module for all SMs.

File: libvcx/src/aries/handlers/proof_presentation/prover/prover.rs (hunk 11 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/proof_presentation/prover/prover.rs:19]: dtto, `state_machine` is a private module for all SMs.

File: libvcx/src/aries/handlers/proof_presentation/verifier/verifier.rs (hunk 13 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/proof_presentation/verifier/verifier.rs:15]: dtto

File: libvcx/src/libindy/utils/wallet.rs (hunk 14 above)
Comment (mirgee) [reply libvcx/src/libindy/utils/wallet.rs:545]: I think the purpose was to remove the only dependence on `api_c` in the file. Not related to the rest of the changes in this PR anyway.

File: wrappers/node/test/suite1/ariesvcx-credential.test.ts (hunk 15 above)
Comment (mirgee) [reply wrappers/node/test/suite1/ariesvcx-credential.test.ts:97]: Did that everywhere in wrapper unit tests

File: wrappers/node/src/api/connection.ts [hunk 17]
```diff
@@ -105,7 +105,7 @@ export interface IConnectionData {
   endpoint: string;
   uuid: string;
   wallet: string;
-  state: StateType;
+  state: number;
```
Comment (Patrik-Stas) [inline wrappers/node/src/api/connection.ts:108]: I think we should return `ConnectionStateType`

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
I think it's important for readibility of the code consuming this wrapper to return Struct/Enums from TS function returning state representation (getState, updateState)

**Review by Patrik-Stas — APPROVED** [review]
